src/Http/Paginators/RequestPaginator.php

```diff
@@ -0,0 +1,281 @@
+// TODO 2: Make it easier to extend the RequestPaginator. Preferably via callbacks, so we ideally don't even need separate classes.
+// TODO 3: Because both page-based and offset-based pagination just bumps the 'paging' number,
+//           would it make sense to do all pagination in _the_ RequestPaginator, but allow to specify the counter somehow?
+//         Maybe a callback that receives a copy of the original Request, as well as the latest Response (which has the corresponding latest Request and PendingRequest),
+//           and have that callback set the next 'page' on the new Request?
+
+/**
+ * @template TRequest of \Saloon\Contracts\Request
+ * @template TResponse of \Saloon\Contracts\Response
+ *
+ * @implements RequestPaginatorContract<TRequest, TResponse>
+ */
+abstract class RequestPaginator implements RequestPaginatorContract, SerialisableRequestPaginator
+{
+    /**
+     * @var bool
+     */
+    protected bool $async = false;
+
+    /**
+     * Whether or not the Paginator will skip rewinding, if it's used in a new loop.
+     * If not, it'll continue where it left off in a previous loop.
+     * While iterators, by design, starts over, Request Paginators should continue.
+     * Otherwise it'll re-send earlier requests.
+     * So make it default not to rewind, unless explicitly told to do so.
+     *
+     * @var bool
+     *
+     * @see \Saloon\Contracts\RequestPaginator::ignoreRewinding()
+     *
+     * @TODO Come up with a better name for this method.
+     */
+    protected bool $shouldRewind = false;
+
+    /**
+     * @var TResponse|null
+     */
+    protected ?Response $currentResponse = null;
+
+    /**
+     * @param \Saloon\Contracts\Connector $connector
+     * @param TRequest $originalRequest
+     * @param int|null $limit
+     */
+    public function __construct(
+        protected readonly Connector $connector,
+        protected readonly Request $originalRequest,
+        protected ?int $limit = null,
+    ) {}
+
+    /**
+     * Apply paging information, like setting the Request query parameter 'page', or 'offset', etc.
+     *
+     * @param \Saloon\Contracts\Request $request
+     *
+     * @return void
+     */
+    abstract protected function applyPaging(Request $request): void;
+
+    /**
+     * @param callable(int $pendingRequests): (int)|int $concurrency
+     * @param callable(TResponse $response, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $responseHandler
+     * @param callable(mixed $reason, array-key $key, \GuzzleHttp\Promise\PromiseInterface $poolAggregate): (void)|null $exceptionHandler
+     * @return \Saloon\Contracts\Pool
+     */
+    public function pool(
+        callable|int $concurrency = 5,
+        ?callable $responseHandler = null,
+        ?callable $exceptionHandler = null,
+    ): Pool {
+        // The purpose of a pool is to concurrently send requests.
+        // So 'force' set async.
+        $this->async();
```

Comment:
If the iterator provides request classes, then the pool will use `sendAsync`

Reply:
Or is this just so the paginator is also "async"

Reply:
Yeah, exactly. We mark the internal state to be async, so the paginator will create Promises, instead of requests.
It's a bit of WIP, though, as we still need to figure out how to retrieve total pages when we're doing Promises.